test/test_fully_resolve.rb

```diff
@@ -0,0 +1,282 @@
+          "type" => "object",
+          "required" => ["name"],
+          "properties"  => {
+            "name"  => { "type" => "string" }
+          }
+        },
+        {
+          "$schema" => "http://json-schema.org/draft-04/schema#",
+          "type" => "object",
+          "required" => ["age"],
+          "properties"  => {
+            "age"  => { "type" => "integer" }
+          }
+        }
+      ]
+    }
+
+    resolved = JSON::Validator.fully_resolve(schema)
+    assert_equal(resolved, expected, "should resolve a top level $ref reference")
+
+    # Test that the result is usable for validation
+    data = {name: '2'}
+    assert_valid(schema, data)
+    assert_valid(resolved, data)
+
+    data = {age: 2}
+    assert_valid(schema, data)
+    assert_valid(resolved, data)
+  end
+
+  def test_linked_list_ref
+    stub_request(:get, "example.com/linked_list").to_return(:body => File.new('test/schemas/fully_resolve/linked_list_schema.json'), :status => 200)
+
+    schema = {
+      "$schema" => "http://json-schema.org/draft-04/schema#",
+      "$ref" => "http://example.com/linked_list#"
+      # it does not work to reference the schema by path
+      # "$ref" => 'test/schemas/fully_resolve/linked_list_schema.json'
+    }
+
+    expected = {
+      "$schema" => "http://json-schema.org/draft-04/schema#",
+      "id" => "http://example.com/linked_list#",
+      "type" => "object",
+      "required" => ["head", "tail"],
+      "properties"  => {
+        "head"  => {
+          "type" => "object"
+        },
+        "tail" => {
+        }
+      }
+    }
+    expected["properties"]["tail"]["oneOf"] = [
+      expected,
+      { "type" => "null" }
+    ]
+
+    resolved = JSON::Validator.fully_resolve(schema)
+    assert_equal(resolved, expected, "should resolve a top level $ref reference")
+
+    data = {"head" => {}, "tail" => nil}
+    assert_valid(schema, data)
+    # you don't want to do that it causes infinite recursion aka SystemStackError: stack level too deep
+    #assert_valid(resolved, data)
+  end
+
+  def test_resolution_of_draft_4_spec
+    stub_request(:get, "http://json-schema.org/draft-04/schema#").to_return(:body => File.new('resources/draft-04.json'), :status => 200)
+    schema = {
+      "$schema" => "http://json-schema.org/draft-04/schema#",
+      "$ref" => "http://json-schema.org/draft-04/schema#"
+    }
+
+    schema_array = {
+      "type" => "array",
+      "minItems" => 1,
+      "items" => { "$ref" => "#" }
+    }
+    positive_integer = {
+      "type" => "integer",
+      "minimum" => 0
+    }
+    positive_integer_default_0 = {
+      "allOf" => [ positive_integer, { "default" => 0 } ]
+    }
+    string_array = {
+      "type" => "array",
+      "items" => { "type" => "string" },
+      "minItems" => 1,
+      "uniqueItems" => true
+    }
+    simple_types = {
+      "enum" => [ "array", "boolean", "integer", "null", "number", "object", "string" ]
+    }
+    expected = {}
+    expected.merge!({
```

Comment:
I do, because I need a reference to the outer hash within the merged hash which describes the resolved schema as it is self-referencing.